File: java-soa/inventory-service/src/main/java/com/ecommerce/inventory/application/InventoryServiceImpl.java
package com.ecommerce.inventory.application;

import com.ecommerce.inventory.domain.model.InventoryItem;
import com.ecommerce.inventory.domain.service.InventoryService;
import com.ecommerce.inventory.infrastructure.messaging.InventoryEventPublisher;
import com.ecommerce.inventory.infrastructure.persistence.InventoryPersistenceAdapter;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.Optional;
import java.util.UUID;

/**
 * Inventory Service Implementation - Application Layer
 * 
 * <p><b>Saga Pattern - Participant</b>: This service participates in the
 * order saga by handling stock reservations and releases. It publishes
 * events to communicate the result back to the Order Service.</p>
 * 
 * <p><b>Strategy Pattern</b>: The reservation strategy is encapsulated here.
 * Currently uses a simple "first come, first served" approach. Could be
 * extended with priority queues or batch reservation strategies.</p>
 * 
 * <p><b>SOLID - SRP</b>: Orchestrates inventory operations between
 * persistence and messaging layers.</p>
 */
@Service
@Transactional
public class InventoryServiceImpl implements InventoryService {

    private static final Logger log = LoggerFactory.getLogger(InventoryServiceImpl.class);

    private final InventoryPersistenceAdapter persistenceAdapter;
    private final InventoryEventPublisher eventPublisher;

    public InventoryServiceImpl(InventoryPersistenceAdapter persistenceAdapter,
                                 InventoryEventPublisher eventPublisher) {
        this.persistenceAdapter = persistenceAdapter;
        this.eventPublisher = eventPublisher;
    }

    @Override
    public InventoryItem createInventoryItem(UUID productId, String productName, int initialQuantity) {
        log.info("Creating inventory item: productId={}, name={}, qty={}",
                productId, productName, initialQuantity);

        InventoryItem item = InventoryItem.create(productId, productName, initialQuantity);
        InventoryItem saved = persistenceAdapter.save(item);

        log.info("Inventory item created: id={}, productId={}", saved.getId(), saved.getProductId());
        return saved;
    }

    @Override
    @Transactional(readOnly = true)
    public Optional<InventoryItem> getInventoryByProductId(UUID productId) {
        return persistenceAdapter.findByProductId(productId);
    }

    @Override
    @Transactional(readOnly = true)
    public List<InventoryItem> getAllInventory() {
        return persistenceAdapter.findAll();
    }

    /**
     * Reserves stock for an order - Saga forward action.
     * 
     * <p><b>Saga Step 2</b>: Called when an order.created event is received.
     * Attempts to reserve the requested quantity. Publishes either
     * stock.reserved or stock.insufficient based on availability.</p>
     * 
     * @return true if stock was successfully reserved
     */
    @Override
    public boolean reserveStock(UUID productId, int quantity, UUID orderId) {
        log.info("Attempting to reserve stock: productId={}, qty={}, orderId={}",
                productId, quantity, orderId);

        Optional<InventoryItem> optItem = persistenceAdapter.findByProductId(productId);

        if (optItem.isEmpty()) {
            log.warn("Product not found in inventory: productId={}", productId);
            eventPublisher.publishStockInsufficient(orderId, productId, quantity,
                    "Product not found in inventory");
            return false;
        }

        InventoryItem item = optItem.get();

        if (!item.hasAvailableStock(quantity)) {
            log.warn("Insufficient stock: productId={}, available={}, requested={}",
                    productId, item.getQuantityAvailable(), quantity);
            eventPublisher.publishStockInsufficient(orderId, productId, quantity,
                    "Insufficient stock available");
            return false;
        }

        // Reserve the stock
        item.reserveStock(quantity);
        persistenceAdapter.save(item);

        // Record the transaction
        persistenceAdapter.recordTransaction(productId, "RESERVE", quantity, orderId);

        // Publish success event
        eventPublisher.publishStockReserved(orderId, productId, quantity);

        log.info("Stock reserved successfully: productId={}, qty={}, orderId={}",
                productId, quantity, orderId);
        return true;
    }

    /**
     * Releases reserved stock - Saga compensating action.
     * 
     * <p><b>Saga Compensation</b>: Called when an order.cancelled event is received.
     * Releases the previously reserved stock back to available.</p>
     */
    @Override
    public void releaseStock(UUID productId, int quantity, UUID orderId) {
        log.info("Releasing reserved stock: productId={}, qty={}, orderId={}",
                productId, quantity, orderId);

        InventoryItem item = persistenceAdapter.findByProductId(productId)
                .orElseThrow(() -> new RuntimeException(
                        "Inventory item not found for product: " + productId));

        // Release the reservation
        item.releaseStock(quantity);
        persistenceAdapter.save(item);

        // Record the transaction
        persistenceAdapter.recordTransaction(productId, "RELEASE", quantity, orderId);

        // Publish release event
        eventPublisher.publishStockReleased(orderId, productId, quantity);

        log.info("Stock released: productId={}, qty={}, orderId={}", productId, quantity, orderId);
    }

    @Override
    public InventoryItem addStock(UUID productId, int quantity) {
        log.info("Adding stock: productId={}, qty={}", productId, quantity);

        InventoryItem item = persistenceAdapter.findByProductId(productId)
                .orElseThrow(() -> new RuntimeException(
                        "Inventory item not found for product: " + productId));

        item.addStock(quantity);
        InventoryItem saved = persistenceAdapter.save(item);

        // Record the transaction
        persistenceAdapter.recordTransaction(productId, "ADD", quantity, null);

        log.info("Stock added: productId={}, newAvailable={}", productId, saved.getQuantityAvailable());
        return saved;
    }
}

File: dotnet-soa/InventoryService/Application/InventoryServiceImpl.cs
using InventoryService.Domain.Interfaces;
using InventoryService.Domain.Models;
using InventoryService.Infrastructure.Messaging;
using InventoryService.Infrastructure.Persistence;

namespace InventoryService.Application;

/// <summary>
/// Inventory service implementation - Saga Participant.
/// 
/// Design Pattern: Saga Participant - this service participates in the
/// order creation saga by attempting stock reservation and publishing
/// the result back to the ESB for the orchestrator (Order Service).
/// 
/// Saga Participation:
/// 1. Receives "order.created" → attempts stock reservation
/// 2a. Success → publishes "stock.reserved"
/// 2b. Failure → publishes "stock.insufficient"
/// 3. Receives "order.cancelled" → releases reserved stock (compensation)
/// </summary>
public class InventoryServiceImpl : IInventoryService
{
    private readonly InventoryRepository _repository;
    private readonly KafkaProducer _kafkaProducer;
    private readonly ILogger<InventoryServiceImpl> _logger;

    public InventoryServiceImpl(
        InventoryRepository repository,
        KafkaProducer kafkaProducer,
        ILogger<InventoryServiceImpl> logger)
    {
        _repository = repository;
        _kafkaProducer = kafkaProducer;
        _logger = logger;
    }

    /// <inheritdoc />
    public async Task<InventoryItem> CreateOrUpdateInventoryAsync(Guid productId, string productName, int quantity)
    {
        var existing = await _repository.GetByProductIdAsync(productId);

        if (existing is not null)
        {
            return await _repository.UpdateQuantityAsync(productId, existing.Quantity + quantity);
        }

        var item = new InventoryItem
        {
            ProductId = productId,
            ProductName = productName,
            Quantity = quantity,
            ReservedQuantity = 0
        };

        return await _repository.CreateAsync(item);
    }

    /// <inheritdoc />
    public async Task<IEnumerable<InventoryItem>> GetAllInventoryAsync()
    {
        return await _repository.GetAllAsync();
    }

    /// <inheritdoc />
    public async Task<InventoryItem?> GetByProductIdAsync(Guid productId)
    {
        return await _repository.GetByProductIdAsync(productId);
    }

    /// <inheritdoc />
    public async Task ReserveStockAsync(Guid orderId, List<ReservationRequest> items)
    {
        _logger.LogInformation("Attempting stock reservation for order {OrderId}", orderId);

        // Check availability for all items
        foreach (var item in items)
        {
            var inventory = await _repository.GetByProductIdAsync(item.ProductId);

            if (inventory is null || inventory.AvailableQuantity < item.Quantity)
            {
                var reason = inventory is null
                    ? $"Product {item.ProductName} not found in inventory"
                    : $"Insufficient stock for {item.ProductName}. Available: {inventory.AvailableQuantity}, Requested: {item.Quantity}";

                _logger.LogWarning("Stock reservation failed for order {OrderId}: {Reason}", orderId, reason);

                // Publish stock.insufficient event
                await PublishStockInsufficientAsync(orderId, reason);
                return;
            }
        }

        // Reserve stock for all items
        foreach (var item in items)
        {
            await _repository.ReserveStockAsync(item.ProductId, item.Quantity);
        }

        _logger.LogInformation("Stock reserved successfully for order {OrderId}", orderId);

        // Publish stock.reserved event
        try
        {
            await _kafkaProducer.PublishAsync("stock.reserved", orderId.ToString(), new StockReservedEvent
            {
                OrderId = orderId,
                Timestamp = DateTime.UtcNow
            });
            _logger.LogInformation("Published stock.reserved event for order {OrderId}", orderId);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Failed to publish stock.reserved event for order {OrderId}", orderId);
        }
    }

    /// <inheritdoc />
    public async Task ReleaseStockAsync(Guid orderId, List<ReservationRequest> items)
    {
        _logger.LogInformation("Releasing reserved stock for cancelled order {OrderId}", orderId);

        foreach (var item in items)
        {
            await _repository.ReleaseStockAsync(item.ProductId, item.Quantity);
        }

        // Publish stock.released event
        try
        {
            await _kafkaProducer.PublishAsync("stock.released", orderId.ToString(), new StockReleasedEvent
            {
                OrderId = orderId,
                Timestamp = DateTime.UtcNow
            });
            _logger.LogInformation("Published stock.released event for order {OrderId}", orderId);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Failed to publish stock.released event for order {OrderId}", orderId);
        }
    }

    private async Task PublishStockInsufficientAsync(Guid orderId, string reason)
    {
        try
        {
            await _kafkaProducer.PublishAsync("stock.insufficient", orderId.ToString(), new StockInsufficientEvent
            {
                OrderId = orderId,
                Reason = reason,
                Timestamp = DateTime.UtcNow
            });
            _logger.LogInformation("Published stock.insufficient event for order {OrderId}", orderId);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Failed to publish stock.insufficient event for order {OrderId}", orderId);
        }
    }
}
